### PR TITLE
Retry when east-west gateway hostname resolution fails

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
Fixes https://github.com/istio/istio/issues/44155

### CODE REVIEW

File: pilot/pkg/model/network.go [hunk 1]
```diff
@@ -267,7 +269,7 @@ func (mgr *NetworkManager) IsMultiNetworkEnabled() bool {
 	}
 	mgr.mu.RLock()
 	defer mgr.mu.RUnlock()
-	return len(mgr.byNetwork) > 0
+	return mgr.multiNetworkEnabled
```
Comment (howardjohn) [inline pilot/pkg/model/network.go:272]: why do we need this change?

File: pilot/pkg/model/network.go [hunk 2]
```diff
@@ -482,11 +491,19 @@ func (n *networkGatewayNameCache) resolveAndCache(name string) []string {
 		ttl = MinGatewayTTL
 	}
 	expiry := time.Now().Add(ttl)
+	refreshPeriod := ttl
+	if len(addrs) == 0 {
+		// resolution failed, retry after 5s (by default)
+		// ttl is set to ~136 years
```
Comment (howardjohn) [inline pilot/pkg/model/network.go:497]: I am not sure I understand where 136y is from or why

File: pilot/pkg/model/network.go [hunk 3]
```diff
@@ -482,11 +491,19 @@ func (n *networkGatewayNameCache) resolveAndCache(name string) []string {
 		ttl = MinGatewayTTL
 	}
 	expiry := time.Now().Add(ttl)
+	refreshPeriod := ttl
+	if len(addrs) == 0 {
+		// resolution failed, retry after 5s (by default)
```
Comment (howardjohn) [inline pilot/pkg/model/network.go:496]: every 5s forever may be high load. May want backoff? 

30s seems reasonable  for failure if no backoff?

File: pilot/pkg/model/network_test.go [hunk 4]
```diff
@@ -89,9 +90,35 @@ func TestGatewayHostnames(t *testing.T) {
...
+			return !reflect.DeepEqual(env.NetworkManager.AllGateways(), gateways)
 		})
+		xdsUpdater.WaitOrFail(t, "xds full")
+	})
+
+	workingDNSServer.setHosts(make(sets.Set[string]))
+	gateways = env.NetworkManager.AllGateways()
+	t.Run("resolution failed", func(t *testing.T) {
+		// the gateways should remain
+		retry.Until(func() bool {
```
Comment (howardjohn) [inline pilot/pkg/model/network_test.go:102]: retry function should not call Fatal in the function, should use `retryOrFail` if you want to fail.

Right now this ALWAYS runs for 1s which is unexpected. Same with L111

Comment (howardjohn) [reply pilot/pkg/model/network_test.go:102]: Or if we mean to check "remains the same for 1s" we can use other mechanisms, like just `xdsUpdater.AssertEmpty` and then check the AllGateways once. And use a lower time than 1s like 50ms

File: pilot/pkg/model/network.go (hunk 1 above)
Comment (dddddai) [reply pilot/pkg/model/network.go:272]: Please see https://github.com/istio/istio/issues/44155#issue-1645326965
> When resolving addresses, the NetworkManager first [does this](https://github.com/istio/istio/blob/4157e1887071037b369fd82cb4559dc7628bb57b/pilot/pkg/model/network.go#L221), and then tries to [add the gateway info back](https://github.com/istio/istio/blob/4157e1887071037b369fd82cb4559dc7628bb57b/pilot/pkg/model/network.go#L244) once it has resolved the address. However, if the IP address cannot be resolved, the gateway info is not added back. This leaves the NetworkManager in a status where it thinks [multinetwork is disabled](https://github.com/istio/istio/blob/4157e1887071037b369fd82cb4559dc7628bb57b/pilot/pkg/model/network.go#L270), causing stiod to [push all endpoints](https://github.com/istio/istio/blob/4157e1887071037b369fd82cb4559dc7628bb57b/pilot/pkg/xds/ep_filters.go#L39-L42) to the sidecar.

File: pilot/pkg/model/network.go (hunk 2 above)
Comment (dddddai) [reply pilot/pkg/model/network.go:497]: It's the _default_ TTL
https://github.com/istio/istio/blob/1.17.1/pilot/pkg/model/network.go#L528
So if resolution failed, the TTL here will be 4294967295/86400/365 ~= 136

File: pilot/pkg/model/network.go (hunk 3 above)
Comment (dddddai) [reply pilot/pkg/model/network.go:496]: I just use the default DNS refresh rate in Envoy

During the downtime we just send DNS requests every 5s without pushing xDS, so I think it's fine 
IMO 30s or backoff may be a bit too long for the system to recover

File: pilot/pkg/model/network_test.go (hunk 4 above)
Comment (dddddai) [reply pilot/pkg/model/network_test.go]: Thank you, fixed

File: pilot/pkg/model/network.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/network.go:272]: What does it send when it failed to resolve hostname with this patch?

Comment (hzxuzhonghu) [reply pilot/pkg/model/network.go:272]: If the resolve never succeed, then what does it send?

Comment (dddddai) [reply pilot/pkg/model/network.go:272]: Then it only will only send endpoints in the same network, without ew-gateways
https://github.com/istio/istio/blob/1.17.1/pilot/pkg/xds/ep_filters.go#L117

File: pilot/pkg/model/network.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/model/network.go:496]: This is 17280 requests per day per hostname per istiod, if its constantly in a failure state, can be very high cost

Comment (dddddai) [reply pilot/pkg/model/network.go]: Got it, changed to 30s

File: pilot/pkg/model/network.go [hunk 5]
```diff
@@ -233,7 +234,7 @@ func (mgr *NetworkManager) resolveHostnameGateways(gatewaySet map[NetworkGateway
 	for host, addrs := range mgr.NameCache.Resolve(names) {
 		gwsForHost := hostnameGateways[host]
 		if len(addrs) == 0 {
```
Comment (howardjohn) [inline pilot/pkg/model/network.go:236]: sorry I didn't notice t his on the first review - len(addrs)==0 is NOT an error IMO. We should distinguish from actual errors fetching an address and getting a succesful response with 0 IP addresses. IMO, in the later case, its not appropriate to retain the last know values

Caveat: I am not a DNS expert. Maybe we can check how Envoy STRICT_DNS cluster works and see if they agree with what I said

**Review by howardjohn — COMMENTED** [review]
general LGTM, 1 comment

File: pilot/pkg/model/network.go (hunk 5 above)
Comment (dddddai) [reply pilot/pkg/model/network.go:239]: Fixed
ref: https://github.com/envoyproxy/envoy/issues/20890

File: pilot/pkg/model/network.go [hunk 6]
```diff
@@ -472,21 +483,32 @@ func (n *networkGatewayNameCache) resolveFromCache(name string) []string {
...
 	// avoid excessive pushes due to small TTL
 	if ttl < MinGatewayTTL {
 		ttl = MinGatewayTTL
 	}
 	expiry := time.Now().Add(ttl)
+	refreshPeriod := ttl
+	if len(addrs) == 0 {
+		// no answer, retry after 30s (by default)
+		// ttl is set to ~136 years
+		refreshPeriod = DNSRetryPeriod
```
Comment (howardjohn) [inline pilot/pkg/model/network.go:501]: For addrs=0, the DNS server can still return a valid TTL I think. If not, we would already default it to MinGatewayTTL which s the same as DNSRetryPeriod. So I think we should remove this, to allow the DNS server to set a TTL >30s when there are 0 IPs.

File: pilot/pkg/model/network.go [hunk 7]
```diff
@@ -510,22 +532,22 @@ func (n *networkGatewayNameCache) refreshAndNotify(name string) func() {
...
 
 		res := n.client.Query(new(dns.Msg).SetQuestion(dns.Fqdn(name), dnsType))
-		if len(res.Answer) == 0 {
-			return
-		}
 
 		mu.Lock()
 		defer mu.Unlock()
+		if res.Rcode == dns.RcodeServerFailure {
+			errs = multierror.Append(errs, fmt.Errorf("upstream dns failure, qtype: %v", dnsType))
```
Comment (howardjohn) [inline pilot/pkg/model/network.go:549]: Appending in concurrent go-routines is not safe

File: pilot/pkg/model/network.go [hunk 8]
```diff
@@ -545,7 +567,11 @@ func (n *networkGatewayNameCache) resolve(name string) ([]string, time.Duration)
 	wg.Wait()
 
 	sort.Strings(out)
-	return out, time.Duration(ttl) * time.Second
+	if errs.Len() == 2 {
+		// return error only if all requests are failed
+		err = errs
```
Comment (howardjohn) [inline pilot/pkg/model/network.go:572]: nit: IMO using the named return vals is confusing, just make this `return out, ttl, nil`

**Review by howardjohn — COMMENTED** [review]
Sorry for all the back and forth. This is really close, just concerned about the non-threadsafe multierror can cause crashes

File: pilot/pkg/model/network.go (hunk 7 above)
Comment (dddddai) [reply pilot/pkg/model/network.go:549]: >Appending in concurrent go-routines is not safe

This is protected by the mutex

File: pilot/pkg/model/network.go (hunk 6 above)
Comment (dddddai) [reply pilot/pkg/model/network.go:501]: > For addrs=0, the DNS server can still return a valid TTL I think

I don't think so, TTL is for each answer(address)

> If not, we would already default it to MinGatewayTTL which s the same as DNSRetryPeriod

If len(addrs)==0, the TTL will be 136y, why would it be `MinGatewayTTL`?

Comment (howardjohn) [reply pilot/pkg/model/network.go:501]: A dns server can return a ttl for an empty response I believe

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/model/network.go (hunk 6 above)
Comment (dddddai) [reply pilot/pkg/model/network.go:501]: Oh yes, sorry I was mixed up, let me fix this

Comment (dddddai) [reply pilot/pkg/model/network.go:501]: Please remove hold if it looks good to you🙂

**Review by howardjohn — APPROVED** [review]

**Pull request merged**